data-plane/src/server/data_plane_server.rs

```diff
@@ -54,88 +56,353 @@ where
+                if read_bytes == 0 {
+                    break;
+                }
+
+                buffer.extend_from_slice(&request_chunk[..read_bytes]);
+
+                if read_bytes < request_chunk.len() {
+                    break;
+                }
+            }
+            Err(e) => {
+                eprintln!("Error reading from stream: {}", e);
+                return Err(crate::error::Error::Io(e));
+            }
+        }
+    }
+    Ok(buffer)
+}
+
+async fn pipe_to_customer_process<L>(
+    stream: &mut TlsStream<L>,
+    buffer: &[u8],
+    port: u16,
+) -> Result<()>
+where
+    TlsStream<L>: AsyncReadExt + Unpin + AsyncWrite,
+{
+    let mut customer_stream = TcpStream::connect(("127.0.0.1", port)).await.unwrap();
+    customer_stream.write_all(buffer).await.unwrap();
+    pipe_streams(stream, customer_stream).await?;
+    Ok(())
+}
+
+fn build_http_request(
+    request: httparse::Request<'_, '_>,
+    body_buffer: &[u8],
+    port: u16,
+) -> Result<Request<Body>> {
+    let uri = format!("http://127.0.0.1:{}{}", port, request.path.unwrap());
+    let mut header_map = HeaderMap::new();
+    for header in request.headers {
+        header_map.insert(
+            HeaderName::from_str(header.name)?,
+            HeaderValue::from_bytes(header.value)?,
+        );
+    }
+
+    let mut req = Request::builder()
+        .uri(uri)
+        .method(request.method.unwrap())
+        .body(Body::from(
+            String::from_utf8(body_buffer.to_vec()).expect("Found invalid UTF-8"),
+        ))?;
+    *req.headers_mut() = header_map;
+    Ok(req)
+}
+
+async fn response_to_bytes(response: Response<Body>) -> Vec<u8> {
+    let mut bytes = Vec::new();
+
+    // TODO: handle other versions
+    let status_line = format!(
+        "HTTP/{} {} {}\r\n",
+        "1.1",
+        response.status().as_u16(),
+        response.status().canonical_reason().unwrap_or("")
+    );
+    bytes.extend_from_slice(status_line.as_bytes());
+
+    for (header_name, header_value) in response.headers() {
+        let header_str = format!(
+            "{}: {}\r\n",
+            header_name.as_str(),
+            header_value.to_str().unwrap_or("")
+        );
+        bytes.extend_from_slice(header_str.as_bytes());
+    }
+
+    bytes.extend_from_slice(b"\r\n");
+
+    let body_bytes: Bytes = hyper::body::to_bytes(response.into_body())
+        .await
+        .unwrap_or_else(|_| Bytes::new());
+
+    bytes.extend_from_slice(&body_bytes);
+
+    bytes
+}
+
+async fn handle_http_request(
+    mut req: Request<Body>,
+    e3_client_for_tcp: Arc<E3Client>,
+    tx_for_tcp: UnboundedSender<LogHandlerMessage>,
+    remote_ip: Option<String>,
+    port: u16,
+) -> Result<Response<Body>> {
+    let e3_client_for_req = e3_client_for_tcp.clone();
+    let feature_context = FeatureContext::get();
+    let cage_context = CageContext::get()?;
+    let tx_for_req = tx_for_tcp.clone();
+    let remote_ip = remote_ip.clone();
+    async move {
```

Comment:
How come this is an async move closure?

Reply:
copy and paste during refactor, removed it there